hathor/sysctl/protocol.py

```diff
@@ -13,85 +13,66 @@
 # limitations under the License.
 
 import inspect
-import json
-from typing import TYPE_CHECKING, Any, Callable, Optional
+from typing import Callable, Optional
 
 from pydantic import ValidationError
 from twisted.protocols.basic import LineReceiver
 
-from hathor.sysctl.exception import SysctlEntryNotFound, SysctlException, SysctlReadOnlyEntry, SysctlWriteOnlyEntry
-
-if TYPE_CHECKING:
-    from hathor.sysctl.sysctl import Sysctl
+from hathor.sysctl.exception import (
+    SysctlEntryNotFound,
+    SysctlException,
+    SysctlReadOnlyEntry,
+    SysctlRunnerException,
+    SysctlWriteOnlyEntry,
+)
+from hathor.sysctl.runner import SysctlRunner
 
 
 class SysctlProtocol(LineReceiver):
     delimiter = b'\n'
 
-    def __init__(self, root: 'Sysctl') -> None:
-        self.root = root
+    def __init__(self, runner: SysctlRunner) -> None:
+        self.runner = runner
 
     def lineReceived(self, raw: bytes) -> None:
         try:
             line = raw.decode('utf-8').strip()
         except UnicodeDecodeError:
             self.sendError('command is not utf-8 valid')
+
         if line.startswith('!help'):
             _, _, path = line.partition(' ')
             self.help(path)
             return
-        elif line == '!backup':
+        elif line.startswith('!backup'):
             self.backup()
             return
-        head, separator, tail = line.partition('=')
-        head = head.strip()
-        tail = tail.strip()
-        if separator == '=':
-            self.set(head, tail)
-        else:
-            self.get(head)
-
-    def sendError(self, msg: str) -> None:
-        """Send an error message to the client. Used when a command fails."""
-        self.sendLine(f'[error] {msg}'.encode('utf-8'))
 
-    def set(self, path: str, value_str: str) -> None:
-        """Run a `set` command in sysctl."""
         try:
-            value = self._deserialize(value_str)
-        except json.JSONDecodeError:
-            self.sendError('value: wrong format')
-            return
-
-        try:
-            self.root.set(path, value)
-        except SysctlEntryNotFound:
-            self.sendError(f'{path} not found')
-        except SysctlReadOnlyEntry:
-            self.sendError(f'cannot write to {path}')
+            feedback = self.runner.run(line)
+            if feedback:
+                self.sendLine(feedback)
+        except SysctlEntryNotFound as e:
+            self.sendError(f'{str(e)} not found')
```

Comment:
Why not use `line` here? Or ask the path to the runner?

Reply:
We can't ask the path to the runner because the runner is a singleton and don't hold any line state. We can use the `line` here, however it will come with the assigned value to the path as well. Maybe the following exceptions should be in the runner side. There we can put the path in the exception message.

```python
        except SysctlEntryNotFound as e:
            self.sendError(f'{str(e)} not found')
        except SysctlReadOnlyEntry as e:
            self.sendError(f'cannot write to {str(e)}')
        except SysctlWriteOnlyEntry as e:
            self.sendError(f'cannot read from {str(e)}')
```

Reply:
But the runner knows how to parse the line and provide the key. So we can move it to separate method and use it here.

Reply:
Oh! Ok, I will see this possibility.

Reply:
Done: dc44366f1b24bfc91eeae13e63ee36f0cc29e0c0

Reply:
Thanks!